src/inmanta/agent/agent.py

```diff
@@ -334,6 +336,78 @@ def notify(
             self.future.set_result(ResourceActionResult(cancel=False))
 
 
+@dataclasses.dataclass
+class DeployRequest:
+    """
+    A request to perform a deploy
+
+    :param is_repair: is this a full deploy/repair run or incremental?
+    :param is_periodic: is this deploy triggered by a timer?
+    :param reason: textual description of the deployment
+    """
+
+    is_repair: bool
+    is_periodic: bool
+    reason: str
+
+    def interrupt(self, other: "DeployRequest") -> "DeployRequest":
+        """Interrupt this deploy and for the other and produce a new request"""
+        return DeployRequest(
+            self.is_repair, self.is_periodic, "Restarting run '%s', interrupted for '%s'" % (self.reason, other.reason)
+        )
+
+
+class DeployRequestAction(str, enum.Enum):
+    """
+    When a deploy is running and a new request arrives, we can take the following actions
+    """
+
+    ignore = "ignore"
+    """ ignore the new request, continue with the exiting deploy """
+    terminate = "terminate"
+    """ terminate the current deploy, start the new one """
+    defer = "defer"
+    """ defer the new request,  continue with the exiting deploy, start the new one when the current deploy is done"""
+    interrupt = "interrupt"
+    """ interrupt the current deploy: cancel it, start the new one and restart the current deploy when the new one is done"""
+
+
+# This matrix describes what do when a new DeployRequest enters before the old one is done
+# Format is old_is_repair, old_is_periodic, new_is_repair, new_is_periodic
+# The underlying idea is that
+# 1. periodic deploys have no time pressure, they can be delayed
+# 2. non-periodic deploy should run as soon as possible
+# 3. non-periodic incremental deploy take precedence over repairs (as they are smaller)
+# 4. periodic deploys should not interrupt each other to prevent restart loops
+deploy_response_matrix = {
```

Comment:
I found this very difficult to interpret. I had to convert it to conditions to be able to see the full picture. If I understand correctly you started out with a conditional implementation and then converted it to this matrix. Do you find it more readable? What if we split up the conditions for which takes precedence and whether or not the non-preferred needs to be run afterwards? e.g. (bad naming and docstrings, some more cleanup needed)
```python
    def take_action(self, other: "DeployRequest") -> tuple["DeployRequestAction", Optional["DeployRequest"]]:
        """
        Returns the action to take and the deploy request to schedule afterwards, if appliccable
        """
        carry_on: bool = self.pick(other)
        if not carry_on:
            if self.run_after(other):
                return DeployRequestAction.interrupt, self.interrupt(other)
            else:
                return DeployRequestAction.terminate, None
        if other.run_after(self):
            return DeployRequestAction.defer, other
        else:
            return DeployRequestAction.ignore, None

    def pick(self, other: "DeployRequest") -> bool:
        """
        Return True iff self should be preferred over other.
        """
        if self.is_periodic:
            return (
                # if both are periodic, pick the full one
                self.is_repair
                if (
                    other.is_periodic
                    # if neither is full, pick the new one
                    and self.is_repair != other.is_repair
                )
                # otherwise prefer non-periodic
                else False
            )
        elif other.is_periodic:
            # periodic may only interrupt full if both are incremental
            return not self.is_repair and not other.is_repair
        elif self.is_repair != other.is_repair:
            # neither is periodic: incremental takes precedence over full
            return not self.is_repair
        else:
            # pick newest when both are incremental or both are full
            return False

    def run_after(self, other: "DeployRequest") -> bool:
        """
        Should self still run after other has finished?
        """
        # TODO: this is tightly coupled to the logic in `pick`
        return (
            # incremental can always be skipped
            self.is_repair
            # no need to run if other was already a full deploy
            and not other.is_repair
        )
```

Reply:
I think this approach has some advantages:

1. it is clear that all cases are covered. 
2. it is clear how each case is covered 
3. some properties are very easy to see (e.g. the jobs that can be resumed are all full repairs)

What drove me to this is that the 4 level deep if statement allows for some very nasty bugs, like confusing `is_repair` with `self.is_repair`. There are so many code paths that contain similar code that testing against mixups is very hard. Tests must traverse all branches.  Here, there is basically one code path, so every mistake is systemic.

Reply:
Perhaps we have to compact the tuples to 

`(periodic_full, increment)`
or
`(PF,I)`

Reply:
For 1-3 we could have a test case that builds the full matrix and asserts it is equal to the one you built by hand (I meant to include that suggestion in my earlier comment but it seems I forgot). In my opinion that would give us the best of both worlds: code with readable decision drivers, and a matrix for verifying specific cases.

Reply:
I also don't find the branching code particularly understandable (the table is not very understandable either), so readability is not a strong argument to me.